query_optimizer/PhysicalGenerator.cpp

```diff
@@ -176,6 +177,8 @@ P::PhysicalPtr PhysicalGenerator::optimizePlan() {
     rules.emplace_back(new AttachLIPFilters());
   }
 
+  rules.push_back(std::make_unique<Partition>(optimizer_context_));
```

Comment:
I have revisited the partition rule, and think it is ok for now to put it between two LIP-related rules, as for `HashJoin` we only have the following changes to a physical plan
 1. add additional `Selections` for base tables or subquery relations, and a few non-LIP related operators like `sort` and `union all`.
 1. change output `PartitionSchemeHeader` of a Physical Plan node.

But for partitioned aggregation in a follow-up PR, we may add another aggregate operator as a finalize step. On the other hand, the original aggregation maybe substituted by a newly created aggregation with the same input node, so I put the partition rule before `AttachLIPFilters`.